### ISSUE TITLE
Arena durable game recordを既存GUIで再生できる牌譜Viewerを追加する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-play#26]
Parent architecture decision: https://github.com/lisbun/lisjong-project/issues/7
Live viewer predecessor: #25
Durable record prerequisite: https://github.com/lisbun/lisjong-arena/issues/155 — **COMPLETE / PR #194 merged 2026-09-09**

# Status

**READY / UNBLOCKED — Arena #155 durable local game record v1 is merged. Implementation may start against the supported strict loader; #25 presentation reuse remains preferred where available but is not a hard prerequisite.**

Arena #155 / PR #194 now provides the concrete persisted source this Issue was waiting for:

```text
completed standard RiichiEnv LocalGameRunner
    ↓
Arena-owned durable record schema v1
    + objective GameTrace
    + LocalGameResult
    + step/seat PolicyInput + DecisionTrace
    + exact provenance / payload digests / deterministic record identity
    ↓
strict cross-process loader
```

The implementation is intentionally not a project-wide canonical `GameRecord` and not a viewer schema. This Issue should consume the supported strict loader rather than parse the raw bundle independently.

`lisjong-play` の既存Tkinter GUIと #25 のlive Spectator modeを人間向けpresentationとして再利用し、Arena #155で保存されたsuccessful local gameのdurable recordをprocess終了後に読み込み、**牌譜としてGUIで再生・移動・観戦できるReplay Viewer**を追加する。

本Issueの目的はgeneric replay engineやcanonical `GameRecord`を作ることではない。

```text
Arena #155 durable record
        ↓
Arena-supported strict loader
        ↓
Replay presentation source
        ↓
shared lisjong-play board/result presentation
        ↓
human viewer
```

を成立させる。

---

# Goal

operatorが#155のcompleted durable recordを選択し、live gameを再実行せずに、記録された対局を既存GUIで閲覧できるようにする。

最低限:

- 牌譜をopenする
- 対局先頭から表示する
- 前後のpresentation/action boundaryへ移動する
- 局単位で移動する
- 自動再生 / pause / speed変更
- 局結果を確認する
- 半荘結果 / final rankingを確認する

を可能にする。

---

# Source-of-truth boundary

牌譜のsource of truthはArena #155の**strict-loaded completed durable record**とする。

`lisjong-play` がraw filesを独自にbest-effort parseしない。

```text
supported loader PASS
    ↓
Replay Viewer

unsupported / corrupt / truncated / tampered
    ↓
fail closed
```

#155のmanifest / identity / digest / completion semanticsを迂回しない。

牌譜を表示するためにrecordを書き換えたり、欠損情報を推測で補完したりしない。

---

# Presentation reuse

新しい牌譜専用盤面rendererをゼロから作らない。

conceptual target:

```text
Human live source ──────┐
                        │
Spectator live source ──┼─> shared board/result presentation
                        │
Replay source ──────────┘
```

#25が先に実装済みならそこで成立したshared presentationを再利用する。

#25が未完了の場合でも、Replay要件を理由にproject-wide `ViewerState` frameworkを先行設計せず、current GUIの実際の共通presentation部分だけを再利用・抽出する。

---

# Replay boundary semantics

Viewer navigationの単位を曖昧にしない。

実装前に#155 recordが保持する:

- `GameTrace`
- step ordinals
- GameTrace event sequence intervals
- decision observations
- round / final result metadata

を確認し、**recorded authoritative boundaryから一意に定義できる最小のpresentation step**をlockする。

初期候補はcompleted applied step / objective-event boundaryである。

GUI都合で複数eventsを麻雀ruleに基づき勝手に再解釈し、「1打牌」「1巡」等のcanonical unitを作らない。

---

# No rule re-execution in Viewer

`lisjong-play` Replay Viewerは以下を再実装しない。

- legal action generation
- call priority
- score calculation
- yaku / fu
- riichi settlement
- round progression rules
- hidden-hand inference
- shanten / ukeire

recorded objective events / engine-owned presentation facts / supported replay projectionから表示状態を構成する。

もし#155のdurable recordだけでは盤面表示に必要な状態を安全・一意に再構成できず、`lisjong-play`側で麻雀ruleを再実装する必要が生じる場合は、その場で推測実装せず:

```text
REPLAY PROJECTION PREREQUISITE REQUIRED
```

としてArenaまたはengine側のbounded prerequisiteを提案する。

---

# Hidden information / spectator presentation

牌譜Viewerでは観戦目的として、recordに正当に含まれたconcealed informationを表示してよい。

ただし:

- recorded `PolicyInput`をomniscient stateへ書き換えない
- player-safe decision snapshotとobjective/offline informationをsemantic上区別する
- hidden informationをPolicy実行へ戻さない
- Viewer内でtraining ground truth / HandBelief labelを作らない

を維持する。

4席のconcealed handを時系列で表示するために必要なauthoritative dataがrecordに不足する場合は、ambiguous reconstructionで埋めない。

---

# Navigation controls

最低限次を提供する。

```text
First
Previous
Next
Round Start / previous round boundary
Round End / next round boundary
Play
Pause
Speed
```

exact naming / placement / keyboard shortcutはcurrent GUI conventionsへ合わせる。

## Backward navigation

backward moveのためにengineを逆実行しない。

recorded boundaryからdeterministically presentation stateを再構成するか、Replay source内でbounded snapshots/indexを持つ。

GUI widget stateそのものをsource of truthにしない。

## Random access

初期Issueではarbitrary decision search / timeline scrubber / database indexを必須にしない。

局boundaryと前後step移動が成立すればよい。

---

# Result / metadata presentation

最低限、recorded provenanceから人間が識別できる範囲で:

- seed
- game mode
- Policy assignment
- record identity / schema version (詳細画面やlogでもよい)
- 局情報
- scores
- 局結果
- 最終rank

を確認可能にする。

machine-facing digestを常時盤面へ大量表示する必要はない。

#155のmetadataをUI向けに表示名へ変換しても、semantic identityを独自再定義しない。

---

# Relationship to AI analysis

#155にDecisionTrace / AnalysisTraceが存在しても、初期牌譜ViewerではAI reasoning表示をcompletion requirementにしない。

ただし将来:

```text
analysis result
  ↓ seed/game/decision reference
Replay Viewer
  ↓ jump to recorded decision
```

へ拡張できるよう、recorded step / decision identityとの対応を失わない。

Q値・logits・HandBelief・Mortal comparison等をこのIssueで追加しない。

---

# File / opening UX

minimum UXとしてoperatorがcompleted record bundle/pathを選択できること。

Tkinterのfile/directory chooser等、current desktop GUIに自然な方法を使ってよい。

recent-file database、library catalog、thumbnail、record management UIは作らない。

unsupported recordのエラーは人間が理解できる形で表示するが、loader failureを成功扱いしない。

---

# Threading / playback lifecycle

- Tk widgetはmain threadだけで操作する
- auto-play timer / background readbackがTk thread ruleを破らない
- replayはlive engine workerを不要に起動しない
- pause / next / previousを高速操作してもstate raceを起こさない
- closeでtimer / background taskを解放する

Replay Viewerはrecorded gameを再実行してPolicyを呼ばない。

---

# Preflight

実装開始時に最低限確認する。

- this Issue / latest comments
- `lisbun/lisjong-project#7` final handoff
- `lisjong-play #25` current implementation / result
- Arena #155 / merged PR #194 implementation / exact durable schema v1 / strict loader API
- `lisjong-play` current main / `AGENTS.md`
- `README.md`
- `docs/architecture.md`
- current GUI board / result renderer
- #25で抽出済みのshared presentation boundary, if any
- #155のGameTrace / step / decision / final result readback semantics
- concealed hand / board reconstructionに必要なauthoritative record coverage

不足情報を`lisjong-play`のprivate rule/state reconstructionで補わない。

---

# Tests

## Input / loader

- valid #155 completed recordをopenできる
- unknown schema versionをrejectする
- corrupt / truncated / digest mismatch recordをrejectする
- invalid recordをpartial replayとして表示しない

## Navigation

- First / Next / Previousがdeterministic
- round boundary navigationがrecorded round identityと一致
- forward -> backward -> forwardで同じpresentation stateへ戻る
- end boundaryを越えて進まない
- start boundaryより前へ戻らない
- play / pause / speedがnavigation semanticsを変えない

## Presentation

- live Viewerと主要board/result renderingを共有する
- scores / rivers / melds / riichi / dora / round metadataをrecorded semanticsから表示する
- 局結果 / 半荘結果をrecorded resultから表示する
- concealed handを表示する場合、authoritative record sourceだけを使う

## Non-interference

- replay中にengine / Policyを実行しない
- Viewerでlegality / scoring / progressionを再計算しない
- recordをmutationしない

large corpusはCIへ含めず、小さいfixture / focused persisted recordと最小integrationを使う。

---

# Documentation

最低限更新候補:

- `README.md`
- `docs/architecture.md`

明記する:

```text
live Human Play
live AI Spectator      #25
persisted Replay       this Issue
raw durable record     Arena #155 / COMPLETE
```

Replay ViewerはArena #155のconsumerであり、record schema ownerではないことを明記する。

---

# Non-goals

- durable record schemaのowner移動
- #155 raw formatの独自fork
- generic replay engine
- project-wide canonical GameRecord
- Viewer-side mahjong rule engine
- record repair / best-effort recovery
- record editor
- save/resume live game
- database / record library
- cloud sync
- web viewer
- RiichiLab / Tenhou / Mortal arbitrary log import
- video export
- arbitrary timeline scrubber
- AI reasoning / logits / Q / HandBelief visualization
- analysis-to-viewer deep link実装
- training dataset generation

---

# Acceptance criteria

- [x] Arena #155 durable record prerequisite completed / merged
- [ ] Arena #155のstrict durable record/readback contractをpreflightした
- [ ] #155 unsupported/corrupt recordをfail closedにする
- [ ] live gameを再実行せずrecordをopenできる
- [ ] existing/#25 board/result presentationを再利用している
- [ ] first / previous / next / round-boundary navigationが成立する
- [ ] auto-play / pause / speed changeが成立する
- [ ] backward navigationでengine逆実行やGUI widget state依存をしていない
- [ ] scores / rivers / melds / riichi / dora / round progressをrecorded semanticsから表示できる
- [ ] 局結果を表示できる
- [ ] 半荘結果 / final rankingを表示できる
- [ ] concealed handsを表示する場合、recorded authoritative dataだけを使用する
- [ ] player-safe decision inputとoffline/objective informationを混同していない
- [ ] Viewer側でrule / scoring / legality / progressionを再実装していない
- [ ] recordをmutationしていない
- [ ] replay中にPolicyを実行していない
- [ ] Tk lifecycle / timer / close semanticsが安全
- [ ] affected tests / docsを更新している
- [ ] generic replay framework / database / analysis UIへscopeを広げていない

# Current next action

```text
implementation-ready

1. preflight merged Arena #155 / PR #194 strict loader and record coverage
2. inspect current lisjong-play presentation boundary / #25 state
3. if record coverage is sufficient, implement the smallest Replay source + navigation layer
4. if authoritative presentation state cannot be reconstructed without Mahjong rule reimplementation, stop and create a bounded Arena/engine replay-projection prerequisite
```

# Expected future follow-up

Small Experiment Loop / diagnosticsがmatureした後、experiment resultのseed/game/decision referenceから該当局面をこのReplay Viewerで直接開くbounded integrationを別Issueとして検討する。

**Comment by lisbun** [comment]
## Implementation handoff — Arena durable record v2 / PR #208 merged

Replay Viewer本体は PR #33 でmerge済み。Arena側のbounded prerequisiteとして `lisbun/lisjong-arena#207` を進め、PR #208 で durable local game record **schema v2** がmergeされた。

Arena v2では、既存の:

```text
objective GameTrace
+ LocalGameResult
+ step/seat PolicyInput + DecisionTrace
```

に加えて、strict loaderからtypedな:

```text
LocalGameInspection.round_results
```

を取得できる。

各 `RoundResult` は少なくとも:

```text
round identity
start/end scores
riichi sticks before/after
dora indicators
riichi seats
wins[]
  winner_seat
  tsumo / loser_seat
  deltas
  ura_indicators
  scoring | None

draw | None
  reason
  exhaustive
  deltas
```

を保持する。

`wins[].scoring` は RiichiEnv 0.4.8 の backend state がcapture可能な局だけ存在し、非最終局では多くの場合 `None`。これは意図したgapであり、Viewer側で yaku / han / fu / points を推測・再計算して埋めてはいけない。

Arena strict loaderはschema v2で `round_results.json` をdigest / byte count / record identityへ含め、objective `GameTrace`とのsame-run bindingも検証する。`start_kyoku` / `hora` / `ryukyoku` / dora / reach_accepted / non-final settlement のrecorded objective factが一致しないrecordはfail closedする。

### Current next implementation

`lisjong-play`側では新しいviewer frameworkを作らず、PR #33のReplay source / controller / shared board rendererをそのまま拡張する。

1. `lisjong-arena` dependency pinを、PR #208 merge後のrevisionへ更新する
2. Replay sourceをArena v2 strict loader / typed `inspection.round_results` 前提へ更新する
3. 既存のraw `GameTrace` terminal parsingによる局結果表示を、typed `RoundResult` consumerへ置き換える
4. 局結果表示に、authoritativeに得られる範囲で以下を追加する
   - winner / ron・tsumo / loser
   - score deltas / start-end scores
   - dora / ura indicators
   - draw reason / exhaustive distinction
   - `scoring is not None` の場合のみ backend-computed han / fu / yaku / payment facts
5. `scoring is None` の局は「詳細取得不可」として扱い、delta等からscore/yaku/fuを逆算しない
6. Replay Viewerは引き続きRiichiEnv / scorer / HandEvaluator / tenpai evaluatorをimport・実行しない
7. v1 recordはArena v2 loaderがrejectする現在のcontractに従い、Play側でbest-effort互換parseしない
8. record mutation、Policy execution、engine rerunを追加しない

### Scope lock

今回のfollow-upに含めない:

```text
RiichiEnv upstream gapの解消
非最終局scoringの再構築
winning tile / winning handの推測
exhaustive draw tenpai推定
riichi declaration river marker
4席concealed handの合成
AI reasoning / logits / Q / HandBelief表示
generic replay framework
```

### Validation

最低限:

- Arena v2 fixture / strict-loaded recordをopenできる
- typed `RoundResult`から局結果をdeterministicに表示する
- scoringあり / scoringなし / drawをそれぞれcoverする
- malformed/tampered recordはArena loader rejectionをそのままfail closedで扱う
- result projectionでMahjong rule codeを呼ばない
- navigation / autoplay / shared board renderingの既存testを壊さない

### Completion judgement

このfollow-up merge後、#26 Acceptance Criteriaを再確認する。

Arena #207自体には RiichiEnv 0.4.8 upstream gap が残るが、#26側は「recorded authoritative dataだけを表示し、不足は推測しない」というIssue contractに従ってcompletion可否を判断する。

**Comment by lisbun** [comment]
## Status update — Replay Viewer implementation is complete on current `main`

The remaining v2 follow-up has now landed:

- PR #33 — Replay Viewer / navigation / shared GUI presentation
- `lisjong-arena` PR #208 — durable local game record schema v2 with typed per-round result facts
- PR #34 — consume Arena v2 typed `round_results` in the Viewer
  - merged as `63782809c782bf7a1a19a4ba12db9521d7b4fc3d`

The Viewer now reads round results only through Arena's supported strict loader and typed `LocalGameInspection.round_results`. The old `lisjong-play`-side raw MJAI terminal-event parsing has been removed.

Current supported presentation includes, where authoritatively recorded:

- round identity / scores / riichi-stick settlement
- dora indicators / riichi seats
- winner / tsumo-ron / deal-in seat
- score deltas
- exhaustive-vs-abortive draw distinction
- backend-computed han / fu / yaku / payments / pao when `RoundWinScoring` is available
- ura indicators only for a recorded riichi winner
- final scores / ranking

When backend scoring is unavailable, the Viewer explicitly shows that it was not recorded and does **not** reconstruct it from deltas or other state.

Known source-side gaps remain and are intentionally **not** guessed in the Viewer:

- winning tile / winning hand
- non-final-round backend scoring under RiichiEnv 0.4.8
- exhaustive-draw tenpai seats
- riichi declaration river marker
- authoritative four-seat concealed-hand history

Those are not outstanding Viewer implementation defects. The bounded Arena-side result-capture work and its current RiichiEnv limitations are documented in `lisjong-arena#207` / PR #208.

### Current completion judgement

For this Issue's original Replay Viewer scope, the acceptance criteria are now satisfied on `main`: persisted replay opens through the strict loader, navigation/autoplay work without engine re-execution, existing GUI presentation is reused, recorded round/final results are displayed, and missing information is left unavailable rather than reconstructed.

If richer replay truth becomes available from a future RiichiEnv contract, it should be added as a new bounded enrichment follow-up rather than reopening Viewer-side rule reconstruction.
